opts/options.go

```diff
@@ -190,6 +227,17 @@ func Protocols(protocols ...protocol.ID) Option {
 	}
 }
 
+// ClientProtocols sets the protocols the DHT uses to initiate connections and queries
+//
+// Defaults to dht.DefaultClientProtocols. If Protocols have been set the to the non-default protocols,
+// then ClientProtocols defaults to match Protocols.
+func ClientProtocols(protocols ...protocol.ID) Option {
```

Comment:
Currently, one can specify an arbitrary set of protocols on which to listen when constructing a DHT.

Unfortunately, this was a bit shortsighted. We now need to be able to actually distinguish between two protocols so specifying a "bag of protocols" isn't useful because the protocols matter.

Proposal: require that the DHT protocol always have the form `/CUSTOM_PREFIX/kad/X.Y.Z` where `/CUSTOM_PREFIX/` is the part the user specifies:

```go
func ProtocolPrefix(myPrefix string) Option { ... }
```

That way, we can keep adding support for new protocols seamlessly.

---

Background:

@aschmahmann and I were concerned that this kind of restriction could cause issues for live networks and discussed allowing an arbitrary map (custom protocol -> "canonical" protocol and/or handler).

After some more [research](https://grep.app/search?q=%5C.Protocols%5C%28%5B%5E%29%5D%2B%5C%29&regexp=true&filter[lang][0]=Go), I believe that restricting to a custom prefix will actually be fine:

* Projects that use a custom prefix:
  * Lotus
  * Gcache
  * IPFS
  * OpenBazaar (except testnet which we can probably break).
* Projects that use an _actual_ custom protocol:
  * Prysm (custom protocol for eth2, unreleased)
  * Filecoin (can probably be changed)

Therefore, if we act now, we can avoid this extra complexity.

Reply:
cc @raulk, @whyrusleeping

Reply:
This seems like solid rationale and design proposal.

The reason users use this feature is to namespace their network and avoid DHT merges. A protocol prefix accomplishes that just fine.

It also enables us to iterate on the granularity of the protocol, going forward, e.g. if we want to break down the protocol in finer-grained ones, or add negotiated capabilities like a dedicated "chaperoning" protocol (for clients to beacon presence to servers), reputation, routing table membership signalling, etc., we can do that seamlessly.

In a nutshell, +1 on this proposed approach for extensibility.

P.S.: Did we end up exposing an option for users to enable/disable specific versions of the protocol?

Reply:
Note: the filecoin spec change has been approved https://github.com/filecoin-project/specs/pull/890

Reply:
> P.S.: Did we end up exposing an option for users to enable/disable specific versions of the protocol?

We've already exposed a "use this set of protocol strings" option. We'll have to remove that.

Reply:
> P.S.: Did we end up exposing an option for users to enable/disable specific versions of the protocol?

This PR was meant to deal with separating client and server protocols to support backwards compatible DHT upgrades. By taking @Stebalien 's suggestion to use protocol prefixes (like we do in Bitswap), none of this needs to be exposed to the user. Either a node is in "client" mode and supports kad/2.0.0 or it's in server mode and supports kad/2.0.0, but can also respond to kad/1.0.0 queries (although no kad/1 peers will end up in the routing table of a kad/2 server).